File: content/snippets/go/design_patterns_factory_method.go
package main

import "fmt"

type Parser interface{ Parse() string }

type JSONParser struct{}

func (JSONParser) Parse() string { return "parsed json" }

type CSVParser struct{}

func (CSVParser) Parse() string { return "parsed csv" }

type ImportJob interface {
	Run() string
	createParser() Parser
}

type JSONImportJob struct{}

func (JSONImportJob) createParser() Parser { return JSONParser{} }
func (j JSONImportJob) Run() string        { return j.createParser().Parse() }

type CSVImportJob struct{}

func (CSVImportJob) createParser() Parser { return CSVParser{} }
func (j CSVImportJob) Run() string        { return j.createParser().Parse() }

func main() {
	jobs := []ImportJob{JSONImportJob{}, CSVImportJob{}}
	for _, job := range jobs {
		fmt.Println(job.Run())
	}
}

File: content/snippets/typescript/design_patterns_factory_method.ts
interface Parser {
  parse(): string;
}

class JsonParser implements Parser {
  parse(): string {
    return 'parsed json';
  }
}

class CsvParser implements Parser {
  parse(): string {
    return 'parsed csv';
  }
}

abstract class ImportJob {
  run(): string {
    return this.createParser().parse();
  }

  protected abstract createParser(): Parser;
}

class JsonImportJob extends ImportJob {
  protected createParser(): Parser {
    return new JsonParser();
  }
}

class CsvImportJob extends ImportJob {
  protected createParser(): Parser {
    return new CsvParser();
  }
}

for (const job of [new JsonImportJob(), new CsvImportJob()]) {
  console.log(job.run());
}